chapters/ch15/2.multimodal-and-hybrid-search.ipynb

```diff
@@ -0,0 +1,890 @@
+    "import torch\n",
+    "import clip\n",
+    "from io import BytesIO\n",
+    "\n",
+    "device = \"cuda\" if torch.cuda.is_available() else \"cpu\"\n",
+    "model, preprocess = clip.load(\"ViT-B/32\", device=device)\n",
+    "\n",
+    "engine = get_engine()\n",
+    "spark = SparkSession.builder.appName(\"AIPS\").getOrCreate()"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": 2,
+   "metadata": {},
+   "outputs": [
+    {
+     "name": "stdout",
+     "output_type": "stream",
+     "text": [
+      "Already up to date.\n",
+      "movies_with_image_embeddings.pickle\n"
+     ]
+    }
+   ],
+   "source": [
+    "![ ! -d 'tmdb' ] && git clone --depth 1 https://github.com/ai-powered-search/tmdb.git\n",
+    "! cd tmdb && git pull\n",
+    "! cd tmdb && mkdir -p '../../../data/tmdb/' && tar -xvf movies_with_image_embeddings.tgz -C '../../../data/tmdb/'"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": 3,
+   "metadata": {},
+   "outputs": [
+    {
+     "name": "stdout",
+     "output_type": "stream",
+     "text": [
+      "Wiping \"tmdb\" collection\n",
+      "Creating \"tmdb\" collection\n",
+      "Status: Success\n",
+      "Adding LTR QParser for tmdb collection\n",
+      "Adding LTR Doc Transformer for tmdb collection\n",
+      "../../data/judgments.tgz already exists\n",
+      "../../data/movies.tgz already exists\n",
+      "Successfully written 65616 documents\n"
+     ]
+    }
+   ],
+   "source": [
+    "%run ../ch10/1.setup-the-movie-db.ipynb"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "## Listing 15.14"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": 4,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "from aips.spark import create_view_from_collection\n",
+    "from aips.spark.dataframe import from_sql\n",
+    "from pyspark.sql.window import Window\n",
+    "from pyspark.sql.functions import col, split, regexp_replace\n",
+    "\n",
+    "def normalize_embedding(embedding):\n",
+    "    return numpy.divide(embedding,\n",
+    "      numpy.linalg.norm(embedding,axis=0)).tolist()\n",
+    "\n",
+    "def read(cache_name):\n",
+    "    cache_file_name = f\"../../data/tmdb/{cache_name}.pickle\"\n",
+    "    with open(cache_file_name, \"rb\") as fd:\n",
+    "        return pickle.load(fd)\n",
+    "\n",
+    "def tmdb_with_embeddings_dataframe():\n",
+    "    movies = read(\"movies_with_image_embeddings\")\n",
+    "    embeddings = movies[\"image_embeddings\"]\n",
+    "    normalized_embeddings = [normalize_embedding(e) for e in embeddings]\n",
+    "    movies_dataframe = spark.createDataFrame(\n",
+    "        zip(movies[\"movie_ids\"], movies[\"titles\"], \n",
+    "            movies[\"image_ids\"], normalized_embeddings),\n",
+    "        schema=[\"movie_id\", \"title\", \"image_id\", \"image_embedding\"])\n",
+    "    return movies_dataframe\n",
+    "\n",
+    "def tmdb_lexical_embeddings_dataframe():\n",
```

Comment:
Note: I reverted the `tmdb_with_embeddings_dataframe` changes because they completely changed the dataset coming back (to posters it seems) and so the examples stopped working as expected.

I'm fine with refactoring from SQL to Python, but I need to dataset to remain stable and the image dataset that was there before works much better for the examples than posters.

Reply:
Ticket created